tests/test_utils.py

```diff
@@ -145,18 +145,23 @@ def testAllPairs(self):
         n = 25
         pairs = all_pairs(n)
         nchoose2 = n * (n - 1) // 2
+        # Build all pairs using a loop to ensure numpy upper_triu() ordering matches.
+        pairs_from_loop = [[i, j] for i in range(n) for j in range(n) if i < j]
```

Comment:
do you need to use n?

`[[i, j] for j in range(n) for i in range(j)]` ? (did not check)

Reply:
That gives the same set but a different ordering.

Reply:
yeah I didn't check the ordering, by my point was we don't need both to be `n` i think... then that means we dont need the branch.

Reply:
How's this: 

`[[i, j] for i in range(n - 1) for j in range(i + 1, n)]`?

Reply:
how we went to n-1?

Reply:
for n=3 I thought my code was correct, but I was rushing.

Reply:
you are right, my suggestion is not correct for higher values.  Okay, let's just leave it so the unit test is less confusing.  My bad.

Reply:
the `if i < j` was forcing i to be in range n-1.